safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,218 @@
+
+use std::{path::PathBuf};
+
+use tokio::fs;
+use tokio::runtime::{Builder, Runtime};
+
+use lazy_static::lazy_static;
+use anyhow::{Result, ensure};
+
+use tokio::sync::watch::Receiver;
+use utils::{
+    lsn::Lsn,
+    zid::{ZTenantTimelineId},
+};
+
+use tracing::*;
+
+use crate::remote_storage::{LocalFs, RemoteStorage};
+use crate::{broker, SafeKeeperConf};
+
+lazy_static! {
+    static ref BACKUP_RUNTIME: Runtime = {
+        Builder::new_multi_thread()
+            .worker_threads(16)
+            .enable_all()
+            .build()
+            .unwrap()
+    };
+}
+
+const BACKUP_ELECTION_PATH: &str = "WAL_BACKUP";
+
+// TODO: there's need to track last upload time to avoid losing partially filled segments
+// Situation: compute wrote a log record and went idle (log was backed up to S3)
+// SK got replaced, compute work up wrote log recrods enough to fill up the segment, 
+// we are going to upload the segment but writes before replacement will be missing
+
+#[derive(Clone)]
+pub struct WalBackup {
+    
+    // TODO: Can't figure out how to properly store this
+    _remote_storage : &LocalFs,
+ 
+    conf: SafeKeeperConf,
+
+    backup_lsn : Lsn,
+
+    // etcd lease id, used for leader election
+    lease_id : Option<i64>,
+    timeline_id: ZTenantTimelineId, 
+}
+
+async fn detect_task(
```

Comment:
Can add some `.instrument(info_span!(...` calls around the main sync loop: we can add fields into this span, such as timeline_id to automatically log every entry inside the span with this id.

This way, we can distinguish sync attempts in the logs.

Reply:
made a lousy attempt at this (at the detect task scheduling) but somehow can't log timeline there (something with macro expansion isn't right)